Detectors/CPV/reconstruction/src/FullCluster.cxx

```diff
@@ -0,0 +1,210 @@
+  float x = 0., z = 0.;
+  Geometry::absIdToRelPosInModule(digitIndex, x, z);
+
+  mElementList.emplace_back(digitIndex, energy, x, z, label);
+  mEnergy += energy; //To be updated when calculate cluster properties.
+  mMulDigit++;
+}
+//____________________________________________________________________________
+void FullCluster::evalAll()
+{
+  // Evaluate cluster parameters
+  evalLocalPosition();
+  //  evalGlobalPosition();
+}
+//____________________________________________________________________________
+void FullCluster::purify()
+{
+  // Removes digits below threshold
+
+  float threshold = o2::cpv::CPVSimParams::Instance().mDigitMinEnergy;
+
+  std::vector<CluElement>::iterator itEl = mElementList.begin();
+  while (itEl != mElementList.end()) {
+    if ((*itEl).energy < threshold) { //very rare case
+      itEl = mElementList.erase(itEl);
+    } else {
+      ++itEl;
+    }
+  }
+
+  mMulDigit = mElementList.size();
+
+  if (mMulDigit == 0) { // too soft cluster
+    mEnergy = 0.;
+    return;
+  }
+
+  // Remove non-connected cells
+  if (mMulDigit > 1) {
+    mEnergy = 0.; // Recalculate total energy
+    auto it = mElementList.begin();
+    while (it != mElementList.end()) {
+      bool hasNeighbours = false;
+      for (auto jt = mElementList.begin(); jt != mElementList.end(); ++jt) {
+        if (it == jt) {
+          continue;
+        }
+        if (Geometry::areNeighbours((*it).absId, (*jt).absId) == 1) {
+          hasNeighbours = true;
+          break;
+        }
+      }
+      if (!hasNeighbours) { //Isolated digits are rare
+        it = mElementList.erase(it);
+        --mMulDigit;
+      } else {
+        mEnergy += (*it).energy;
+        ++it;
+      }
+    }
+  } else {
+    mEnergy = mElementList[0].energy;
+  }
+}
+//____________________________________________________________________________
+void FullCluster::evalLocalPosition()
+{
+  // Calculates the center of gravity in the local CPV-module coordinates
+  // Note that correction for non-perpendicular incidence will be applied later
+  // when vertex will be known.
+
+  if (mEnergy <= 0.) { // zero energy cluster, position undefined
+    mLocalPosX = -999.;
+    mLocalPosZ = -999.;
+    return;
+  }
+
+  // find module number
+  mModule = Geometry::absIdToModule(mElementList[0].absId);
+
+  float wtot = 0.;
+  mLocalPosX = 0.;
+  mLocalPosZ = 0.;
+  float invE = 1. / mEnergy;
+  for (auto it : mElementList) {
+    float w = std::max(float(0.), o2::cpv::CPVSimParams::Instance().mLogWeight + std::log(it.energy * invE));
+    mLocalPosX += it.localX * w;
+    mLocalPosZ += it.localZ * w;
+    wtot += w;
+  }
+  if (wtot > 0) {
+    wtot = 1. / wtot;
+    mLocalPosX *= wtot;
+    mLocalPosZ *= wtot;
+  }
+}
+
+//____________________________________________________________________________
+std::vector<Digit>::const_iterator FullCluster::BinarySearch(const std::vector<Digit>* container, Digit& element)
```

Comment:
I don't see where do you use this method, but are your digits sorted (to use binary search)?